### PR TITLE
fix Issue 18708 - Flow analysis in constructors not done correctly fo…

### PR DESCRIPTION
**Pull request by WalterBright**
…r if, ||, &&

### CODE REVIEW

File: src/dmd/ctorflow.d [hunk 1]
```diff
@@ -85,6 +85,22 @@ struct CtorFlow
         foreach (ref u; fieldinit)
             u |= csx;
     }
+
+    /******************************
+     * OR CSX bits to `this`
+     * Params:
+     *  from = bits to OR in
```
Comment (jacob-carlborg) [inline src/dmd/ctorflow.d]: The name of the parameter is `ctorflow`.

File: test/fail_compilation/test18708.d [hunk 2]
```diff
@@ -0,0 +1,64 @@
+/* TEST_OUTPUT:
+---
+fail_compilation/test18708.d(24): Error: one path skips field `s`
```
Comment (jacob-carlborg) [inline test/fail_compilation/test18708.d]: Can the missing path be included in the error message?

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

File: src/dmd/dscope.d [hunk 3]
```diff
@@ -209,21 +209,11 @@ struct Scope
...
-                if (enclosing.ctorflow.fieldinit.length)
-                {
-                    assert(ctorflow.fieldinit.ptr != enclosing.ctorflow.fieldinit.ptr);
-                    foreach (i, u; ctorflow.fieldinit)
-                        enclosing.ctorflow.fieldinit[i] |= u;
-                }
-                ctorflow.freeFieldinit();
-            }
-        }
+            enclosing.ctorflow.OR(ctorflow);
```
Comment (jacob-carlborg) [inline src/dmd/dscope.d]: Indentention looks inconsistent.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

File: src/dmd/dscope.d (hunk 3 above)
Comment (RazvanN7) [reply src/dmd/dscope.d]: It isn't. It looks inconsistent with the deleted lines

Comment (jacob-carlborg) [reply src/dmd/dscope.d]: Ok, I’m a bit limited here on my phone.

File: test/fail_compilation/test18708.d (hunk 2 above)
Comment (WalterBright) [reply test/fail_compilation/test18708.d]: Changing error messages is beyond the scope of this PR.

**Review by jacob-carlborg — APPROVED** [review]

**Pull request merged**